### PR TITLE
Conduit (Part 1)

### PR DESCRIPTION
**Pull request by piegamesde**
###### Motivation for this change

Adds a package, NixOS module and test for [Conduit](conduit.rs). Split up #124435 into two smaller change sets to get things merged more quickly.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [x] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

CC @pstn @andir @mweinelt

### CODE REVIEW

**Comment by pstn** [comment]
What us the reason for this?

**Comment by piegamesde** [comment]
@pstn I'd like to get Conduit merged. I took the parts from the PR that are ready for merge (everything except the nginx options) and made them a separate pull request. You should be able to simply rebase afterwards.

File: nixos/modules/services/misc/matrix-conduit.nix [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+      enable = mkEnableOption "matrix-conduit";
+
+      extraEnvironment = mkOption {
+        type = types.attrsOf types.str;
+        description = "Extra Environment variables to pass to the conduit server.";
+        default = {};
+        example = { RUST_BACKTRACE="yes"; };
+      };
+
+      nginx.enable = mkOption {
```
Comment (mohe2015) [inline nixos/modules/services/misc/matrix-conduit.nix:23]: This options seems to be a no-op and should probably be removed until the intended functionality is discussed and implemented.

File: nixos/tests/matrix-conduit.nix [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+        extraEnvironment.RUST_BACKTRACE = "yes";
+      };
+      services.nginx = {
+        enable = true;
+        virtualHosts.${name} = {
+          enableACME = false;
+          forceSSL = false;
+          enableSSL = false;
+
+          locations."/_matrix" = {
```
Comment (mohe2015) [inline nixos/tests/matrix-conduit.nix:24]: To follow upstream (https://gitlab.com/famedly/conduit/-/blob/next/DEPLOY.md#nginx) and prevent people from copying non-working configuration at least the .well-known config should probably be added.

Comment (mohe2015) [reply nixos/tests/matrix-conduit.nix:24]: According to https://spec.matrix.org/unstable/server-server-api/#resolving-server-names we *should* provide the .well-known path. But as this is a test this is not directly relevant. Also it's only needed for federation. Still we could use this place here to show a good sample configuration e.g. (I hope):
```nix
{
  services.matrix-conduit = {
    enable = true;
    settings = {
      global = {
        server_name = "example.org";
        allow_encryption = true;
        allow_federation = true;
        #allow_registration = true;
      };
    };
  };

  services.nginx = {
    enable = true;
    recommendedGzipSettings = true;
    recommendedOptimisation = true;
    recommendedProxySettings = true;
    recommendedTlsSettings = true;

    virtualHosts."example.org" = {
      enableACME = true;
      forceSSL = true;

      listen = [
        { ssl = true; port = 443; addr = "0.0.0.0"; }
        { ssl = true; port = 443; addr = "[::0]"; }
        { ssl = true; port = 8448; addr = "0.0.0.0"; }
        { ssl = true; port = 8448; addr = "[::0]"; }
      ];

      extraConfig = ''
        merge_slashes off;
      '';

      locations."= /.well-known/matrix/server".extraConfig = ''
        add_header Content-Type application/json;
        return 200 '${builtins.toJSON { "m.server" = "example.org:8448"; }}';
      '';

      locations."/_matrix" = {
        proxyPass = "http://[::1]:6167$request_uri";
        extraConfig = ''
          proxy_buffering off;
        '';
      };
    };
  };

  networking.firewall.allowedTCPPorts = [ 443 80 8448 ];
}
```

File: nixos/modules/services/misc/matrix-conduit.nix (hunk 1 above)
Comment (piegamesde) [reply nixos/modules/services/misc/matrix-conduit.nix:23]: Oh yeah I removed all the options but forgot that one :facepalm:

File: nixos/tests/matrix-conduit.nix (hunk 2 above)
Comment (piegamesde) [reply nixos/tests/matrix-conduit.nix:24]: I don't think tests are a good source of example configuration. It should be either ingrained into the module system or provided in the general documentation (manual, wiki, ...).

File: nixos/modules/services/misc/matrix-conduit.nix [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+              description = ''
+                Whether this server federates with other servers.
+              '';
+            };
+            global.trusted_servers = mkOption {
+              type = types.listOf types.str;
+              default = [ "matrix.org" ];
+              description = "Servers trusted with signing server keys.";
+            };
+
```
Comment (mohe2015) [inline nixos/modules/services/misc/matrix-conduit.nix:74]: nit: empty line

File: nixos/modules/services/misc/matrix-conduit.nix [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+      systemd.services.conduit = {
+        description = "Conduit Matrix Server";
+        documentation = [ "https://gitlab.com/famedly/conduit/" ];
+        wantedBy = [ "multi-user.target" ];
+        environment = lib.mkMerge ([
+          { CONDUIT_CONFIG = configFile; }
+          cfg.extraEnvironment
+        ]);
+        serviceConfig = {
+          DynamicUser = true;
```
Comment (mohe2015) [inline nixos/modules/services/misc/matrix-conduit.nix:113]: don't know if we could restrict the permissions even further. Unfortunately upstream seems to have no recommendation.

**Review by mohe2015 — COMMENTED** [review]
Would really love a section in the manual though, maybe just with my sample config? Or an adaption of it?

File: nixos/modules/services/misc/matrix-conduit.nix (hunk 4 above)
Comment (piegamesde) [reply nixos/modules/services/misc/matrix-conduit.nix:113]: I think I've gone as far as possible. If you know something else to try out, let me know. Upstreams don't usually have much knowledge about Systemd harderning, unfortunately.

**Comment by mohe2015** [comment]
For future work: I think it would be awesome if there would be a test that tests the federation between different matrix servers. So you can specify a list of servers (even mixing different implementations) then a script will create a room for users from each of them and then it's tested whether the federation works properly.

File: nixos/modules/services/misc/matrix-conduit.nix (hunk 4 above)
Comment (mohe2015) [reply nixos/modules/services/misc/matrix-conduit.nix:111]: I think everything except the system call filter (which doesn't work yet) could be done.
```diff
diff --git a/nixos/modules/services/misc/matrix-conduit.nix b/nixos/modules/services/misc/matrix-conduit.nix
index 3f62f3aea87..69622f0381e 100644
--- a/nixos/modules/services/misc/matrix-conduit.nix
+++ b/nixos/modules/services/misc/matrix-conduit.nix
@@ -129,24 +129,31 @@ in
           User = "conduit";
           LockPersonality = true;
           MemoryDenyWriteExecute = true;
+          CapabilityBoundingSet = "";
+          DevicePolicy = "strict";
+          DeviceAllow = [ "/dev/urandom" "/dev/random" "/dev/stdin" "/dev/stdout" "/dev/stderr" ];
           ProtectClock = true;
           ProtectControlGroups = true;
+          ProtectHome = true;
           ProtectHostname = true;
           ProtectKernelLogs = true;
           ProtectKernelModules = true;
           ProtectKernelTunables = true;
+          ProtectProc = "invisible";
           PrivateDevices = true;
           PrivateMounts = true;
           PrivateUsers = true;
+          ProcSubset = "pid";
           RestrictAddressFamilies = [ "AF_INET" "AF_INET6" ];
           RestrictNamespaces = true;
           RestrictRealtime = true;
           SystemCallArchitectures = "native";
           SystemCallFilter = [
-            "@system-service"
-            "~@privileged"
+            # inspiration: systemd-analyze syscall-filter
+            "@default @basic-io @chown @file-system @io-event @network-io @process @signal @sync mprotect"
           ];
           StateDirectory = "matrix-conduit";
+          UMask = "0077";
           ExecStart = "${cfg.package}/bin/conduit";
           Restart = "on-failure";
           RestartSec = 10;

```

Comment (piegamesde) [reply nixos/modules/services/misc/matrix-conduit.nix:111]: Thanks for the diff. I suggest we put this on the backlog and open that topic in a new PR once this one is merged.

**Review by mohe2015 — APPROVED** [review]
LGTM! Have been running this for a few days.

File: pkgs/servers/matrix-conduit/default.nix [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+
+  meta = with lib; {
+    description = "A Matrix homeserver written in Rust";
+    homepage = "https://conduit.rs/";
+    license = licenses.asl20;
+    maintainers = with maintainers; [ pstn piegames ];
+  };
+}
+
+
```
Comment (SuperSandro2000) [inline pkgs/servers/matrix-conduit/default.nix:24]: 2 new lines to much.

File: nixos/tests/matrix-conduit.nix [hunk 6]
```diff
@@ -0,0 +1,98 @@
+import ./make-test-python.nix ({ pkgs, ... } : let
+
+
+  name = "conduit";
+
+in {
+
+
+  nodes = {
```
Comment (SuperSandro2000) [inline nixos/tests/matrix-conduit.nix:9]:
```suggestion
import ./make-test-python.nix ({ pkgs, ... } : let

  name = "conduit";

in {
  nodes = {
```

**Comment by piegamesde** [comment]
@SuperSandro2000 While you're technically correct with your suggestions, I've already asked you multiple times not to review any formatting changes anymore. This is my last polite warning, and I would not like to have to escalate this.

**Comment by worldofgeese** [comment]
As this branch is still unmerged, how would one make use of it to deploy Conduit today on a NixOS installation? I'm a newer Nix person, apologies.

**Pull request merged**